include/xtensor/xfunction.hpp

```diff
@@ -160,8 +160,42 @@ namespace xt
 
         template <class B, class R>
         using functor_batch_simd_type_t = typename functor_batch_simd_type<B, R>::type;
+
+        // State storage
+        template<class S, class is_trivial>
+        struct xfunction_cache_impl
+        {
+            mutable S m_shape;
+            mutable bool m_shape_trivial;
+            mutable bool m_shape_computed;
```

Comment:
I think it would be better to remove the ` mutable` here, and use a global `mutable` for the cache member.

Reply:
If it becomes a `m_cache` we could also use the occasion to rename `m_shape_computed` to something like `is_initialized`, so that it reads `m_cache.is_initialized`